{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "34218fbb",
   "metadata": {},
   "source": [
    "# Análise bioinformática da proteína da cápside codificada pelo GeneID 1258751"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "544e8f1d",
   "metadata": {},
   "source": [
    "## Importação de bibliotecas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b291ff3",
   "metadata": {},
   "source": [
    "Nesta secção são importadas as bibliotecas necessárias para a análise bioinformática nomeadamente módulos do BioPython para acesso às bases de dados do NCBI, manipulação de sequências, alinhamentos múltiplos e construção de árvores filogenéticas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "65c50ce7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO, AlignIO, Phylo\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor\n",
    "\n",
    "Entrez.email = \"dmartins0080@gmail.com\"\n",
    "gene_id = \"1258751\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7e8a741",
   "metadata": {},
   "source": [
    "## Obtenção da sequência de aminoácidos"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4474d4c",
   "metadata": {},
   "source": [
    "É utilizada a base de dados NCBI para obter a sequência proteica associada ao GeneID 1258751. A sequência é descarregada em formato FASTA e utilizada como base para as análises subsequentes.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "58314ad7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proteína associada: NP_049787\n",
      "Descrição da proteína: NP_049787.1 major head protein [Escherichia phage T4]\n",
      "Comprimento: 521\n"
     ]
    }
   ],
   "source": [
    "handle = Entrez.efetch(db=\"gene\", id=gene_id, rettype=\"xml\")\n",
    "gene_record = Entrez.read(handle)\n",
    "handle.close()\n",
    "\n",
    "gene = gene_record[0]\n",
    "\n",
    "capside_id = gene[\"Entrezgene_locus\"][0][\"Gene-commentary_products\"][0][\"Gene-commentary_accession\"]\n",
    "print(\"Proteína associada:\", capside_id)\n",
    "\n",
    "handle = Entrez.efetch(db=\"protein\", id=capside_id, rettype=\"fasta\", retmode=\"text\")\n",
    "protein_record = SeqIO.read(handle, \"fasta\")\n",
    "handle.close()\n",
    "\n",
    "print(\"Descrição da proteína:\" , (protein_record.description))\n",
    "print(\"Comprimento:\", len(protein_record.seq))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19c94f47",
   "metadata": {},
   "source": [
    "## Realização de um BLASTp\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "423875ee",
   "metadata": {},
   "source": [
    "A sequência proteica obtida é analisada por BLASTp contra a base de dados nr (non-redundant), com o objetivo de confirmar a sua identidade funcional e identificar proteínas homólogas. Os 10 melhores alinhamentos são utilizados nos passos seguintes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f218ed4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A correr BLASTp (isto pode demorar)...\n",
      "BLAST terminado. Resultado guardado em gene1258751_blast.xml\n"
     ]
    }
   ],
   "source": [
    "SeqIO.write(protein_record,\"gene1258751_protein.fasta\",\"fasta\")\n",
    "with open(\"gene1258751_protein.fasta\") as seq_fasta:\n",
    "    fasta_seq = seq_fasta.read()\n",
    "\n",
    "result_handle = NCBIWWW.qblast(program=\"blastp\", database=\"nr\", sequence=fasta_seq)\n",
    "\n",
    "with open(\"gene1258751_blast.xml\", \"w\") as out:\n",
    "    out.write(result_handle.read())\n",
    "\n",
    "with open(\"gene1258751_blast.xml\") as handle:\n",
    "    blast_record = NCBIXML.read(handle)\n",
    "\n",
    "print(\"Top 10 hits BLAST:\")\n",
    "for i, alignment in enumerate(blast_record.alignments[:10], start=1):\n",
    "    hsp = alignment.hsps[0]\n",
    "    print(f\"{alignment.accession}\")\n",
    "    print(\"   Descrição:\", alignment.hit_def)\n",
    "    print(\"   Identidade:\", hsp.identities, \"/\", hsp.align_length)\n",
    "    print(\"   E-value:\", hsp.expect)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae5e5747",
   "metadata": {},
   "source": [
    "## Alinhamento múltiplo de sequências"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f057288e",
   "metadata": {},
   "source": [
    "As sequências homólogas selecionadas são alinhadas utilizando o programa Clustal Omega, permitindo avaliar a conservação da sequência proteica entre diferentes organismos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b695048",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accessions usados:\n",
      "  NP_049787\n",
      "  WP_371335560\n",
      "  WP_411401423\n",
      "  YP_010074593\n",
      "  UJD20117\n",
      "  YP_010075713\n",
      "  YP_010072737\n",
      "  YP_010066370\n",
      "  EOQ0467934\n",
      "  UGO56632\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "with open(\"gene1258751_blast.xml\") as handle:\n",
    "    blast_record = NCBIXML.read(handle)\n",
    "\n",
    "hit_accessions = []\n",
    "records=[]\n",
    "for alignment in blast_record.alignments:\n",
    "    hit_accessions.append(alignment.accession)\n",
    "    if len(hit_accessions) == 10:\n",
    "        break\n",
    "\n",
    "print(\"Accessions usados:\")\n",
    "for acc in hit_accessions:\n",
    "    print(' ',(acc))\n",
    "\n",
    "for acc in hit_accessions:\n",
    "    handle = Entrez.efetch(db=\"protein\", id=acc,rettype=\"fasta\", retmode=\"text\")\n",
    "    record = SeqIO.read(handle, \"fasta\")\n",
    "    handle.close()\n",
    "    records.append(record)\n",
    "\n",
    "SeqIO.write(records, \"capsid_hits.fasta\", \"fasta\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75d46a04",
   "metadata": {},
   "source": [
    "Corrido no docker: "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4f10300",
   "metadata": {},
   "source": [
    "docker run --rm -v ${PWD}:/data biocontainers/clustalo:v1.2.4-2-deb_cv1 clustalo -i /data/capsid_hits.fasta -o /data/capsid_alignment.aln --force"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8638fbfa",
   "metadata": {},
   "source": [
    "## Construção da árvore filogenética"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3753ca51",
   "metadata": {},
   "source": [
    "A partir do alinhamento múltiplo é construída uma árvore filogenética pelo método Neighbor-Joining, tendo como base uma matriz de disntâncias de identidade, de forma a visualizar as relações evolutivas entre a proteína de estudo e aquelas que foram usadas no alinhamento.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f48731ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  _______ NP_049787.1 major head protein [Esche...\n",
      " |\n",
      " |        ________________ YP_010072737.1 major head protein [Es...\n",
      " |       |\n",
      " |_______|         _______ YP_010075713.1 major head protein [Se...\n",
      " |       |        |\n",
      " |       |________|        _______ EOQ0467934.1 phage capsid protein [Es...\n",
      " |                |  _____|\n",
      " |                | |     | ________ UGO56632.1 putative major capsid prot...\n",
      " |                |_|     ||\n",
      " |                  |      | YP_010066370.1 major head protein [En...\n",
      "_|                  |\n",
      " |                  |_____ YP_010074593.1 major head protein [Es...\n",
      " |\n",
      " | WP_371335560.1 MULTISPECIES: phage ca...\n",
      " |\n",
      " |_______ WP_411401423.1 hypothetical protein [...\n",
      " |\n",
      " |_______ UJD20117.1 major head protein [Escher...\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alinhamento = AlignIO.read(\"capsid_alignment.aln\", \"clustal\")\n",
    "\n",
    "calculator = DistanceCalculator(\"identity\")\n",
    "distancias = calculator.get_distance(alinhamento)\n",
    "\n",
    "id_para_nome = {}\n",
    "\n",
    "for registo in SeqIO.parse(\"capsid_hits.fasta\", \"fasta\"):\n",
    "    id_para_nome[registo.id] = registo.description\n",
    "\n",
    "constructor = DistanceTreeConstructor()\n",
    "arvore = constructor.nj(distancias)\n",
    "\n",
    "for folha in arvore.get_terminals():\n",
    "    if folha.name in id_para_nome:\n",
    "        folha.name = id_para_nome[folha.name]\n",
    "\n",
    "Phylo.draw_ascii(arvore)\n",
    "\n",
    "\n",
    "Phylo.write(arvore, \"arvore_capside_completa.nwk\", \"newick\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biopython",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
